plugins/modem-manager/fu-sahara-loader.c

```diff
@@ -0,0 +1,582 @@
+
+	/* copy const data to mutable GByteArray */
+	bytes = g_byte_array_sized_new(sz);
+	bytes = g_byte_array_append(bytes, data, sz);
+	chunks = fu_chunk_array_mutable_new(bytes->data, bytes->len, 0, 0, self->maxpktsize_out);
+	for (guint i = 0; i < chunks->len; i++) {
+		chk = g_ptr_array_index(chunks, i);
+
+		if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+			g_debug("sending %u bytes", fu_chunk_get_data_sz(chk));
+
+		if (!g_usb_device_bulk_transfer(fu_usb_device_get_dev(self->usb_device),
+						self->ep_out,
+						fu_chunk_get_data_out(chk),
+						fu_chunk_get_data_sz(chk),
+						&actual_len,
+						IO_TIMEOUT_MS,
+						NULL,
+						error)) {
+			g_prefix_error(error, "failed to do bulk transfer (write): ");
+			return FALSE;
+		}
+		if (actual_len != fu_chunk_get_data_sz(chk)) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "only wrote %" G_GSIZE_FORMAT "bytes",
+				    actual_len);
+			return FALSE;
+		}
+		total_sent += actual_len;
+	}
+	if (sz % self->maxpktsize_out == 0) {
+		if (!g_usb_device_bulk_transfer(fu_usb_device_get_dev(self->usb_device),
+						self->ep_out,
+						NULL,
+						0,
+						NULL,
+						IO_TIMEOUT_MS,
+						NULL,
+						error)) {
+			g_prefix_error(error, "failed to do bulk transfer (write): ");
+			return FALSE;
+		}
+		if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+			g_debug("sent zlp");
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_sahara_loader_write_prog(FuSaharaLoader *self, guint32 offset, guint32 length, GBytes *prog, GError **error)
+{
+	gsize sz;
+	const guint8 *data = g_bytes_get_data(prog, &sz);
+
+	g_return_val_if_fail(offset + length <= sz, FALSE);
+
+	g_debug("SENDING --> RAW_DATA: %u bytes (offset = %u, total len = %lu)",
+		length,
+		offset,
+		sz);
+	return fu_sahara_loader_qdl_write(self, &data[offset], length, error);
+}
+
+static gboolean
+fu_sahara_loader_send_packet(FuSaharaLoader *self, GByteArray *pkt, GError **error)
+{
+	guint8 *data = pkt->data;
+	gsize sz = pkt->len;
+
+	fu_sahara_loader_log_packet((struct sahara_packet *)data, "SENDING -->");
+	return fu_sahara_loader_qdl_write(self, data, sz, error);
+}
+
+/* packet composers */
+static GByteArray *
+fu_sahara_loader_compose_reset_packet(void)
+{
+	GByteArray *self;
+	guint32 len = 0x08;
+
+	self = g_byte_array_sized_new(len);
+	g_byte_array_set_size (self, len);
+
+	((struct sahara_packet *)(self->data))->command_id = GUINT32_TO_LE (SAHARA_RESET_ID);
+	((struct sahara_packet *)(self->data))->length = GUINT32_TO_LE (len);
+
+	return self;
+}
+
+static GByteArray *
+fu_sahara_loader_compose_hello_response_packet(FuSaharaMode mode)
+{
+	GByteArray *self;
+	guint32 len = 0x30;
+
+	self = g_byte_array_sized_new(len);
+	g_byte_array_set_size (self, len);
+
+	((struct sahara_packet *)(self->data))->command_id = GUINT32_TO_LE (SAHARA_HELLO_RESPONSE_ID);
+	((struct sahara_packet *)(self->data))->length = GUINT32_TO_LE (len);
```

Comment:
all this casting looks kinda nasty; can you create a struct sahara_packet, assign values and then memcpy_safe into a buffer?